pkg/funky/router.go

```diff
@@ -0,0 +1,182 @@
+///////////////////////////////////////////////////////////////////////
+// Copyright (c) 2018 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+///////////////////////////////////////////////////////////////////////
+package funky
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync"
+
+	"golang.org/x/sync/semaphore"
+)
+
+// FirstPort the starting port number for servers created by a Router
+const FirstPort uint16 = 9000
+
+// Router an interface for delegating function invocations to idle servers
+type Router interface {
+	Delegate(input Message) (*Message, error)
+	Shutdown() error
+}
+
+// DefaultRouter a struct that hold servers that can be delegated to
+type DefaultRouter struct {
+	servers       []Server
+	serverFactory ServerFactory
+	mutex         *sync.Mutex
+	sem           *semaphore.Weighted
+}
+
+// NewRouter constructor for DefaultRouters
+func NewRouter(numServers int, serverFactory ServerFactory) (*DefaultRouter, error) {
+	if numServers < 1 {
+		return nil, IllegalArgumentError("numServers")
+	}
+
+	servers, err := createServers(numServers, serverFactory)
+	if err != nil {
+		return nil, err
+	}
+
+	return &DefaultRouter{
+		servers:       servers,
+		serverFactory: serverFactory,
+		mutex:         &sync.Mutex{},
+		sem:           semaphore.NewWeighted(int64(numServers)),
+	}, nil
+}
+
+// Delegate delegates function invocation to an idle server
+func (r *DefaultRouter) Delegate(input Message) (*Message, error) {
+	server, err := r.findFreeServer()
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		r.releaseServer(server)
+	}()
+
+	var e Error
+	resp, err := server.Invoke(input)
+	if resp != nil {
+		defer resp.Close()
+	}
+
+	logs := Logs{
+		Stdout: server.Stdout(),
+		Stderr: server.Stderr(),
+	}
+
+	if err != nil {
+		switch err.(type) {
+		case TimeoutError:
+			server.Terminate()
+			server, serverErr := r.serverFactory.CreateServer(server.GetPort())
```

Comment:
This `:=` assignment creates a new, case-scoped var `server` which shadows the original one. It does NOT update the original `server` var! 

Better assign to a new var, e.g. `newServer`, then assign it back to `server` with `server = newServer`. 

Also, when the original server has been terminated, but a new one can't be successfully created, make sure that a **bad** server doesn't make it back into the pool!

Reply:
I knew there was a reason I didn't do this Friday, but it slipped my mind this morning.